### ISSUE TITLE
NaN gradients of RelaxedBernoulli samples for p near 0 and 1 in recent TFP versions

### ISSUE THREAD

**Issue by EiffL** [tensorflow/probability#1393]
I'm encountering some numerical issues when evaluating the gradients of a RelaxedBerrnoulli, with parameter p close to either 0 or 1 (close but not super close). Here is a minimal example:
```python
import tensorflow as tf
import tensorflow_probability as tfp

l = 0.99*tf.ones(shape=[],dtype=tf.float32)

with tf.GradientTape() as tape:
  tape.watch(l)
  p = tfd.RelaxedBernoulli(temperature=0.01, probs=l).sample()
dp_dl = tape.gradient(p,l)
print(p, dp_dl)
```
which returns NaN for the gradient in TFP v0.13 and TF 2.5. **Note that I also see the same thing if I directly try to get the gradient of a Transformed Logistic disribution**, so if there is a problem it's probably not at the level of RelaxedBernoulli.

What is interesting is that this used to work perfectly fine in previous TFP version (at least in v0.9). For comparison I made these two notebooks with different TFP versions:
- [working](https://colab.research.google.com/drive/1xkXYtC3ER1z0r25fkUa2pwV_b7OlQBnP?usp=sharing) with TFP=0.9 and TF=2.1
- [not working](https://colab.research.google.com/drive/17dJFNVC2pPniTUAnJC3HPks2AgKonNOA?usp=sharing) with TFP=0.13 and TF=2.5

Looking at the history of both Logistic and  RelaxedBernoulli I really don't see an obvious reason for this change in behavior, so would appreciate any hints/thoughts on what might cause these numerical issues :-)

**Comment by EiffL** [comment]
Ok... so after some additional digging, I've found the culprit.

It looks like this numerical instability of the gradients was introduced here https://github.com/tensorflow/probability/commit/00fcb96333ecbe37e966bc04dc66146fab054c59 at the level of Sigmoid bijector used in the RelaxedBernoulli. This commit was intended to actually provide a more stable implementation of the bijector, but from what I can see it has see, for some reason, this custom `_stable_sigmoid` messes with the gradients.

Here is an example that demonstrates this:
```python
import tensorflow as tf
import tensorflow_probability as tfp
tfd = tfp.distributions
tfb = tfp.bijectors

# This is the current version of the "stable" sigmoid function used in the bijector
# stolen from here: https://github.com/tensorflow/probability/blob/34934d3227928c6a6a6c714b4c707c0d445380d7/tensorflow_probability/python/bijectors/sigmoid.py#L41
def _stable_sigmoid(x):
  """A (more) numerically stable sigmoid than `tf.math.sigmoid`."""
  x = tf.convert_to_tensor(x)
  if x.dtype == tf.float64:
    cutoff = -20
  else:
    cutoff = -9
  return tf.where(x < cutoff, tf.exp(x), tf.math.sigmoid(x))

# Let's define 2 variants of the Sigmoid bijector, changing just based on the 
# actual sigmoid implementation
class VanillaSigmoid(tfp.bijectors.Sigmoid):
  """ Sigmoid bijector based on default tf.sigmoid
  """
  def _forward(self, x):
    if self._is_standard_sigmoid:
      print('Actual value of x in sigmoid', x)
      return tf.sigmoid(x)
    return self.high * tf.sigmoid(x) + self.low * tf.sigmoid(-x)

class CustomSigmoid(tfp.bijectors.Sigmoid):
  """ Sigmoid bijector based on custom _stable_sigmoid
  """
  def _forward(self, x):
    if self._is_standard_sigmoid:
      print('Actual value of x in sigmoid', x)
      return _stable_sigmoid(x)
    return self.high * _stable_sigmoid(x) + self.low * _stable_sigmoid(-x)

# Testing the gradients of a transformed Logistic distribution
p = 0.99*tf.ones(shape=[],dtype=tf.float32)

with tf.GradientTape(persistent=True) as tape:
  tape.watch(p)
  logistic_scale = tf.math.reciprocal(0.01)
  logits_parameter = tf.math.log(p) - tf.math.log1p(-p)
  logistic_loc = logits_parameter * logistic_scale

  # Retrieve a sample from transformed distribution with both sigmoids
  s_vanilla = tfd.TransformedDistribution(tfd.Logistic(logistic_loc, logistic_scale),
                                  VanillaSigmoid()).sample()
  s_custom = tfd.TransformedDistribution(tfd.Logistic(logistic_loc, logistic_scale),
                                  CustomSigmoid()).sample()
  
grad_vanilla = tape.gradient(s_vanilla, p)
grad_custom = tape.gradient(s_custom, p)

print(" ")
print("Gradient using tf.sigmoid", grad_vanilla)
print("Gradient using _stable_sigmoid", grad_custom)
```
returns: 
```
Gradient using tf.sigmoid tf.Tensor(0.0, shape=(), dtype=float32)
Gradient using _stable_sigmoid tf.Tensor(nan, shape=(), dtype=float32)
```

And here is a notebook to try this out live: https://colab.research.google.com/drive/1aeNqgqpJfb7DOd2rs7rGyyrMDVOtq5v2?usp=sharing

Why does the `tf.where` of the "stable" sigmoid breaks the gradients.... I don't know. Because this happens for large values of x here, not at the low value cutoff....

**Comment by srvasude** [comment]
I think we might be able to rollback the changes in that patch, now that the underlying bug is fixed. I'll check that + add these tests to see if that fixes everything.

**Comment by axch** [comment]
Drive-by: If the underlying sigmoid is fixed, we should certainly roll back our workaround.  If it's not, however, then I expect the gradients of `_stable_sigmoid` can be repaired by using double-where:
```
safe_for_exp = tf.where(x < cutoff, x, 0)
safe_for_sigmoid = tf.where(x < cutoff, 0, x)
return tf.where(x < cutoff, tf.exp(safe_for_exp), tf.math.sigmoid(safe_for_sigmoid))
```

For the curious, the underlying problem this fix is trying to resolve is that `tf.where` propagates 0 gradients to the arm not taken.  This is mostly fine, but if `x` is large, the derivative of `exp(x)` will round to `+inf`, and the back-propagation will compute `0 * +inf = nan`.  Then the gradient propagated to `x` will be `<the right answer from the tf.math.sigmoid arm> + <nan from the tf.exp arm> = nan`.  But, `where`-masking the input to `tf.exp` (and `tf.math.sigmoid`, defensively) stops those `nan` gradients.

**Comment by EiffL** [comment]
wow, thanks @axch for the insight! Indeed, using this trick fixes the gradients of the stable_sigmoid. I've updated the notebook linked above to include the comparison with this fix.
